src/ape/managers/chain.py

```diff
@@ -0,0 +1,103 @@
+
+    _networks: NetworkManager
+    _snapshots: List[SnapshotID] = []
+
+    @property
+    def provider(self) -> ProviderAPI:
+        """
+        The active :class:`~ape.api.providers.ProviderAPI`. Most often, this should
+        be a :class:`~ape.api.providers.TestProviderAPI`.
+
+        Raises:
+            :class:`~ape.exceptions.ProviderNotConnectedError`: When not connected
+              to a provider.
+        """
+
+        provider = self._networks.active_provider
+        if not provider:
+            raise ProviderNotConnectedError()
+
+        return provider
+
+    @property
+    def block_number(self) -> int:
+        """
+        The current block number (int).
+        """
+
+        return self.provider.get_block("latest").number
+
+    def snapshot(self) -> SnapshotID:
+        """
+        Record the current state of the blockchain with intent to later
+        call the method :meth:`~ape.managers.chain.ChainManager.revert`
+        to go back to this point. This method is for development networks
+        only.
+
+        Raises:
+            NotImplementedError: When the active provider does not support
+              snapshotting.
+
+        Returns:
+            :class:`~ape.types.SnapshotID`: The snapshot ID.
+        """
+        provider = self._get_test_provider()
+        snapshot_id = provider.snapshot()
+
+        if snapshot_id not in self._snapshots:
+            self._snapshots.append(snapshot_id)
+
+        return snapshot_id
+
+    def restore(self, snapshot_id: Optional[SnapshotID] = None):
+        """
+        Regress the current call using the given snapshot ID.
+        Allows developers to go back to a previous state.
+
+        Raises:
+            NotImplementedError: When the active provider does not support
+              snapshotting.
+            :class:`~ape.exceptions.UnknownSnapshotError`: When the snapshot ID is not cached.
+            :class:`~ape.exceptions.ChainError`: When the snapshot ID does not exist or there are
+              no snapshot IDs to select from.
+
+        Args:
+            snapshot_id (Optional[:class:`~ape.types.SnapshotID`]): The snapshot ID. Defaults
+            to the most recent snapshot ID.
+        """
+        provider = self._get_test_provider()
+        if not self._snapshots:
+            raise ChainError("There are no snapshots to revert to.")
+        elif snapshot_id is None:
+            snapshot_id = self._snapshots.pop()
+        elif snapshot_id not in self._snapshots:
+            raise UnknownSnapshotError(snapshot_id)
+        else:
+            snapshot_index = self._snapshots.index(snapshot_id)
+            self._snapshots = self._snapshots[:snapshot_index]
+
+        provider.revert(snapshot_id)
+
+    def _get_test_provider(self) -> TestProviderAPI:
+        provider = self.provider
+        if not isinstance(provider, TestProviderAPI):
+            raise NotImplementedError(f"Provider '{provider.name}' does not support snapshotting.")
```

Comment:
Hmm, I was picturing letting the lower-level error from `ProviderAPI` bubble up where `NotImplementedError` just tells you right when you use the feature that it's unsupported.

Reply:
`ProviderAPI` does not have development methods though, so it would be like `method not found ` or something.

And `geth` - which _is_ a `TestProviderAPI` (has a `snapshot()` method) will raise it's `NotImplementedError` and that will bubble up normally.

Reply:
The only one someone could get here is if there were using `infura` or something. They'd be having other issues most likely if that were the case anyway. The real `NotImplentedError` _do_ bubble, such as with `geth`.

Reply:
Ah. Well just saying that "chain does not support X feature" will probably grow to more than just snapshots.

Reply:
Yes, this is just a helper method I made, I can make it more subject to change though by allowing a custom message given to it.